File: backend/src/main/java/com/project/service/TrendService.java
package com.project.service;

import com.project.trend.SimpleTrendStrategy;
import com.project.trend.TrendStrategy;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

@Service
public class TrendService {

    private final TrendStrategy trendStrategy;

    public TrendService() {
        // Utilizing the SimpleTrendStrategy we just created
        this.trendStrategy = new SimpleTrendStrategy();
    }

    public List<String> getTrendingWords(Map<String, Integer> oldData, Map<String, Integer> newData) {
        return new ArrayList<>(getTrendingWordFrequencies(oldData, newData).keySet());
    }

    public Map<String, Integer> getTrendingWordFrequencies(Map<String, Integer> oldData, Map<String, Integer> newData) {
        Map<String, Integer> trendingFrequencies = new HashMap<>();

        for (Map.Entry<String, Integer> entry : newData.entrySet()) {
            String word = entry.getKey();
            int newCount = entry.getValue();
            
            // Defaulting old count to 0 if the word is entirely new
            int oldCount = oldData.getOrDefault(word, 0);

            if (trendStrategy.isTrending(oldCount, newCount)) {
                trendingFrequencies.put(word, newCount);
            }
        }

        return trendingFrequencies;
    }

    public Map<String, Integer> getMockOldData() {
        Map<String, Integer> mockData = new HashMap<>();
        mockData.put("ai", 5);
        mockData.put("market", 3);
        mockData.put("war", 2);
        return mockData;
    }
}

File: node-backend/services/trendService.js
/**
 * Core business rules for calculating trending keywords.
 * Mimics Java's SimpleTrendStrategy and TrendService.
 */

/**
 * Returns mock baseline historical word frequency data.
 * @returns {Record<string, number>}
 */
export function getMockOldData() {
  return {
    ai: 5,
    market: 3,
    war: 2
  };
}

/**
 * Trend strategy: True if the current frequency is more than double the old frequency.
 * Matches: newCount > (oldCount * 2)
 *
 * @param {number} oldCount
 * @param {number} newCount
 * @returns {boolean}
 */
export function isTrending(oldCount, newCount) {
  return newCount > (oldCount * 2);
}

/**
 * Filter new word frequencies to only include trending words.
 *
 * @param {Record<string, number>} oldData Historical frequencies.
 * @param {Record<string, number>} newData Current frequencies.
 * @returns {Record<string, number>} Trending frequencies.
 */
export function getTrendingWordFrequencies(oldData = {}, newData = {}) {
  const trendingFrequencies = {};

  for (const [word, newCount] of Object.entries(newData)) {
    const oldCount = oldData[word] || 0;

    if (isTrending(oldCount, newCount)) {
      trendingFrequencies[word] = newCount;
    }
  }

  return trendingFrequencies;
}
